File: cpp/93-restore-ip-addresses/restore-ip-addresses.cpp
class Solution {
public:
    vector<string> res;
    vector<string> path;

    void backtrack(const string &s, int index) {
        if (path.size() == 4) {
            if (index == s.size()) {
                res.push_back(path[0] + "." + path[1] + "." + path[2] + "." + path[3]);
            }
            return;
        }

        for (int len = 1; len <= 3; len++) {
            if (index + len > s.size()) break;

            string part = s.substr(index, len);
            if (part[0] == '0' && part.size() > 1) continue;
            if (stoi(part) > 255) continue;

            path.push_back(part);
            backtrack(s, index + len);
            path.pop_back();
        }
    } 
    vector<string> restoreIpAddresses(string s) {
        backtrack(s, 0);
        return res;
    }
};

File: kotlin/93-restore-ip-addresses/restore-ip-addresses.kt
class Solution {

    fun restoreIpAddresses(s: String): List<String> {

        val result = mutableListOf<String>()
        val parts = mutableListOf<String>()

        backtrack(
            s,
            0,
            parts,
            result
        )

        return result
    }


    private fun backtrack(
        s: String,
        index: Int,
        parts: MutableList<String>,
        result: MutableList<String>
    ) {

        if (index == s.length) {

            result.add(parts.joinToString("."))

            return
        }


        for (end in index + 1..s.length) {

            val segment = s.substring(index, end)

            if (segment.length > 3) {
                continue
            }

            if (segment.toInt() > 255) {
                continue
            }


            parts.add(segment)

            backtrack(
                s,
                end,
                parts,
                result
            )

            if (parts.isNotEmpty()) {
                parts.removeAt(0)
            }
        }
    }
}